### DISCUSSION TITLE
flash internal ExpressLRS module from sd card firmwaare folder

### DISCUSSION THREAD

**Discussion by koolakoff** [EdgeTX/edgetx#3387] [Ideas]
at the moment EdgeTX allow to flash the internal (or external) 4-in-1 module
Steps:
- put the firmware binary to the sd card at 'firmware' folder (i.e file '4in1-1.3.3.7.bin')
- at RC (i.e. tx16s ii) press 'system', go to 'SD-HC CARD', got to 'FIRMWARE'
- press at bin name. from appeared menu may press 'Flash internal Multi' or 'Flash external Multi'

It would be great to have the same functionality for internal (or external) ExpressLRS modules.

In this case you may build i.e. two flash images by ExpressLRS configurator - let say ELRS 2.5 and ELRS 3.1. put them both to tx16s SD card. Then in the field if you need control your SPI-based elrs drone with old ELRS 2.x support - you simply flash tx16s internal elrs module with corresponding image. And if needs switch to newer drone wit elrs 3. - just flash the other image

**Comment by pfeerick** [comment]
I could be wrong, but I have a feeling this is not currently possible as it would require such functionality in ELRS. And the ELRS guys would probably say the workaround for that is to use your phone, and flash over wifi - i.e. save the binaries to your phone, and then use the ELRS hotspot functionality to do the firmware update. But this, and also passthrough flashing of the MPM module (i.e. plug in your transmitter to the computer, and update the MPM directly - like you can already with the internal ELRS on most transmitters) are both features that would be nice to have in the future. 

Actually, scratch that, it should be possible for the internal module - just like with passthrough flashing, I'm not sure about the external one though.

**Reply by koolakoff** [reply]
suppose theoretically/phisically it is possible. In case the EdgeTX can take controll over the 'passthrough hardware'. I know that tx16s uses crossfire protocol to communicate with any internal/external elrs module. I do not know how 'passthrough' is implemented/designed - is it done via crossfire protocol? or it just use some, let say serial bus (I dont' know hw of connection) and if EdgeTX actually can control this interface. But if it can - then it is possible to investigate the ELRS opensource to learn the flashing protocol and just copypaste it... Easy to say, but as an embedded engineer (from very other direction) I know that it may be a really difficult to do...

**Reply by pfeerick** [reply]
As I said at the end ;) 

> Actually, scratch that, it should be possible for the internal module - just like with passthrough flashing, I'm not sure about the external one though.

Updates are done over serial, and the internal module can be power cycled, and held in programming state on power on, whereas I don't think the external modules can be - I don't think the connection to the boot pin is physically there, and there's no software way around that. At the end of the day, these in most cases an ESP8266 or ESP32 (STM32 in some cases), so all the normal rules of programming apply them apply ;)
